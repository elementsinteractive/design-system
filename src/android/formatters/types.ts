export type FormatterConfig = {
  template: string
  header: string
  packageName: string
}
