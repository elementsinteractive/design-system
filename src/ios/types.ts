export type FormatterConfig = {
  template: string
  header: string
  themeName: string
}
