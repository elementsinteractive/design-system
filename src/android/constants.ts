import { join } from 'path'
import { AndroidTheme } from './types'

export const TEMPLATES = (theme: AndroidTheme) => ({
  typographyTemplate: {
    source: join(__dirname, './templates/typographies.ejs'),
    destination: `${theme.typography}/Type.kt`,
    filter: 'isTypography',
    formatter: 'android/compose/typography',
    packageName: theme.typography.packageName,
  },
  spacingsTemplate: {
    source: join(__dirname, './templates/spacings.ejs'),
    destination: `${theme.spacings}/Spacing.kt`,
    filter: 'isSpacing',
    formatter: 'android/spacings',
    packageName: theme.spacings.packageName,
  },
  colorsTemplate: {
    source: join(__dirname, './templates/colors.ejs'),
    destination: `${theme.colors}/Color.kt`,
    filter: 'color',
    formatter: 'android/compose/colors',
    packageName: theme.colors.packageName,
  },
  headerTemplate: join(__dirname, './templates/header.ejs'),
})
