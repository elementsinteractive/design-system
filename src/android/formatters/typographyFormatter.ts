import type { Formatter } from 'style-dictionary'
import { toLower } from 'lodash'

import { FormatterConfig } from './types'

import { useTemplate } from '../../utils'

export const typographyFormatter: (config: FormatterConfig) => Formatter =
  (config: FormatterConfig) =>
  ({ dictionary }) => {
    const { template, header, packageName } = config

    const mapFontWeight = weight => {
      const mapping = {
        Regular: 'Normal',
      }

      return mapping[weight] ?? weight
    }

    const tokens = dictionary.tokens

    const typographies = Object.keys(tokens).map(token => ({
      name: toLower(token),
      ...tokens[token].value,
      fontWeight: mapFontWeight(tokens[token].value.fontWeight),
    }))

    return useTemplate(template)({
      typographies,
      header,
      packageName,
    })
  }
