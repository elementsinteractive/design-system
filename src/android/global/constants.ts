import { join } from 'path'

import { AndroidGlobalOptions } from '../types'

export const ColorTypeFilename = 'ColorTokens.kt'

export const SpacingTypeFilename = 'SpacingTokens.kt'

export const GLOBAL_TEMPLATES = (options: AndroidGlobalOptions) => ({
  colorsTypeTemplate: createColorsTypeTemplate(options),
  spacingsTypeTemplate: createSpacingsTypeTemplate(options),
  headerTemplate: join(__dirname, './templates/header.ejs'),
})

const createSpacingsTypeTemplate = (options: AndroidGlobalOptions) => ({
  source: join(__dirname, './templates/spacingTokens.ejs'),
  destination: `${options.destination}/${SpacingTypeFilename}`,
  filter: 'spacingsType',
  formatter: 'android/spacingsType',
  packageName: options.packageName,
})

const createColorsTypeTemplate = (options: AndroidGlobalOptions) => ({
  source: join(__dirname, './templates/colorTokens.ejs'),
  destination: `${options.destination}/${ColorTypeFilename}`,
  filter: 'colorsType',
  formatter: 'android/compose/colorsType',
  packageName: options.packageName,
})
