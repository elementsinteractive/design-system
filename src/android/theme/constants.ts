import { join } from 'path'
import { AndroidThemeOptions } from '../types'

export const PaletteFilename = 'Palette.kt'
export const SpacingFilename = 'Spacing.kt'
export const TypeFilename = 'Type.kt'
export const ColorTokensFilename = 'ColorTokens.kt'

export const THEME_TEMPLATES = (options: AndroidThemeOptions) => ({
  typographyTemplate: createTypographyTemplate(options),
  spacingsTemplate: createSpacingsTemplate(options),
  paletteTemplate: createPaletteTemplate(options),
  colorTokensTemplate: createColorTokensTemplate(options),
  headerTemplate: join(__dirname, './templates/header.ejs'),
})

const createTypographyTemplate = ({ theme }: AndroidThemeOptions) => ({
  source: join(__dirname, './templates/typographies.ejs'),
  destination: `${theme.typography.destination}/${TypeFilename}`,
  filter: 'typography',
  formatter: 'android/compose/typography',
  packageName: theme.typography.packageName,
})

const createSpacingsTemplate = (options: AndroidThemeOptions) => ({
  source: join(__dirname, './templates/base/spacings.ejs'),
  destination: `${options.theme.spacings.destination}/${SpacingFilename}`,
  filter: 'spacings',
  formatter: 'android/spacings',
  packageName: options.theme.spacings.packageName,
  sharedPackageName: options.globalPackageName,
})

const createPaletteTemplate = ({ theme }: AndroidThemeOptions) => ({
  source: join(__dirname, './templates/shared/palette.ejs'),
  destination: `${theme.palette.destination}/${PaletteFilename}`,
  filter: 'color',
  formatter: 'android/compose/palette',
  packageName: theme.palette.packageName,
  themeName: theme.name,
})

const createColorTokensTemplate = ({ theme }: AndroidThemeOptions) => ({
  source: join(__dirname, './templates/shared/colorTokens.ejs'),
  destination: `${theme.colors.destination}/${ColorTokensFilename}`,
  filter: 'color',
  formatter: 'android/compose/colorTokens',
  packageName: theme.colors.packageName,
  themeName: theme.name,
})
