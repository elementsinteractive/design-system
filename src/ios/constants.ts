import { join } from 'path'

export const TEMPLATES = (destination: string) => ({
  typographyTemplate: {
    source: join(__dirname, './templates/typographies.ejs'),
    destination: `${destination}/Fonts.swift`,
    filter: 'isTypography',
    formatter: 'ios/typography',
  },
  spacingsTemplate: {
    source: join(__dirname, './templates/spacings.ejs'),
    destination: `${destination}/Spacings.swift`,
    filter: 'isSpacing',
    formatter: 'ios/spacings',
  },
  colorsTemplate: {
    source: join(__dirname, './templates/colors.ejs'),
    destination: `${destination}/Colors.swift`,
    filter: 'color',
    formatter: 'ios/colors',
  },
  headerTemplate: join(__dirname, './templates/header.ejs'),
})
