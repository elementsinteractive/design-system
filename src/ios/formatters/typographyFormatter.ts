import Color from 'tinycolor2'
import { useTemplate } from '../../utils'
import type { Dictionary, Formatter } from 'style-dictionary'
import { FormatterConfig } from '../types'

export const typographyFormatter: (config: FormatterConfig) => Formatter =
  (config: FormatterConfig) =>
    ({ dictionary }) => {
      const { template, header, themeName } = config

      const tokens = dictionary.tokens

      const typographies = Object.keys(tokens).map(token => ({
        name: token,
        ...tokens[token].value,
      }))

      return useTemplate(template)({
        typographies,
        header,
        themeName,
      })
    }
