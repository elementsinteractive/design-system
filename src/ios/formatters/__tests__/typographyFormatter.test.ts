import createDictionary from 'style-dictionary/lib/utils/createDictionary'
import createFormatArgs from 'style-dictionary/lib/utils/createFormatArgs'

import { TEMPLATES } from '../../constants'
import { FormatterConfig } from '../../types'
import { typographyFormatter } from '../typographyFormatter'
import { PROPERTIES, HEADER } from '../../../common/formatters/__tests__/constants'

describe('typographyFormatter', () => {
  it('can handle dictionary with typography tokens', () => {
    const { typographyTemplate } = TEMPLATES('')

    const typographyConfig: FormatterConfig = {
      template: typographyTemplate.source,
      header: HEADER,
      themeName: 'LightTheme',
    }

    const dictionary = createDictionary({ properties: PROPERTIES.typography })
    const formatterArgs = createFormatArgs({ dictionary, platform: {} })

    const formatter = typographyFormatter(typographyConfig)(formatterArgs)

    expect(formatter).toMatchSnapshot()
  })
})
