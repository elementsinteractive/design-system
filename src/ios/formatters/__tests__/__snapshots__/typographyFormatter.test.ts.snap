// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`typographyFormatter can handle dictionary with typography tokens 1`] = `
"/* Test header */

import SwiftUI

public extension Font {
  /// Namespace to prevent naming collisions with static accessors on
  /// SwiftUI's Font.
  ///
  /// Xcode's autocomplete allows for easy discovery of design system fonts.
  /// At any call site that requires a font, type \`Font.DesignSystem.<esc>\`
  /// TODO: Implement custom typefaces support.
  /// TODO: Impelment \`paragraphSpacing\`, \`textDecoration\`, and \`textCase\` support.
  struct DesignSystem {
      /// Please, use \`-1.5\` leading with this font.
      public static let H1 = Font.systemFont(
        size: 96,
        weightName: \\"Light\\"
      )
      /// Please, use \`-0.5\` leading with this font.
      public static let H2 = Font.systemFont(
        size: 60,
        weightName: \\"Light\\"
      )
  }

  private static func systemFont(size: CGFloat = 16, weightName: String = \\"Regular\\") -> Font {
      return system(size: size).weight(weightName.asFontWeight())
  }
}

private extension String {
  /// Matches string with \`Font.Weight\` constant.
  func asFontWeight() -> Font.Weight {
      guard let font = Font.Weight.all[lowercased()] else {
          assertionFailure(\\"Unknown font weight name: \\\\(self)\\")
          return .regular
      }
      return font
  }
}

private extension Font.Weight {
  /// Returns name to font weight contstant mapping.
  static var all: [String: Font.Weight] {
      [
          \\"ultralight\\": .ultraLight,
          \\"thin\\": .thin,
          \\"light\\": .light,
          \\"regular\\": .regular,
          \\"medium\\": .medium,
          \\"semibold\\": .semibold,
          \\"bold\\": .bold,
          \\"heavy\\": .heavy,
          \\"black\\": .black
      ]
  }
}"
`;
